File: jdevkit/src/test/java/com/github/klee0kai/android_devkit/collections/weaklist/RefListIteratorTests.java
package com.github.klee0kai.android_devkit.collections.weaklist;

import static com.github.klee0kai.android_devkit.collections.weaklist.ChangeListsHelper.assertListsSame;
import static com.github.klee0kai.android_devkit.collections.weaklist.ChangeListsHelper.changeForEachListsSame;

import org.junit.Test;

import java.util.Arrays;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.ListIterator;

public class RefListIteratorTests {

    @Test(timeout = 100)
    public void removeByIterator() {
        //Given
        List<ChangeListsHelper.SimpleIntPair> originalList = Arrays.asList(
                new ChangeListsHelper.SimpleIntPair(1, 2),
                new ChangeListsHelper.SimpleIntPair(1, 3),
                new ChangeListsHelper.SimpleIntPair(1, 4),
                new ChangeListsHelper.SimpleIntPair(1, 5),
                new ChangeListsHelper.SimpleIntPair(2, 5),
                new ChangeListsHelper.SimpleIntPair(3, 5),
                new ChangeListsHelper.SimpleIntPair(4, 5)
        );
        List<ChangeListsHelper.SimpleIntPair> expList = new LinkedList<>(originalList);
        List<ChangeListsHelper.SimpleIntPair> actList = new SoftList<>(originalList);

        //When
        changeForEachListsSame(Arrays.asList(expList, actList), list -> {
            Iterator<ChangeListsHelper.SimpleIntPair> it = list.iterator();
            while (it.hasNext()) {
                ChangeListsHelper.SimpleIntPair item = it.next();
                if (item.a >= 2)
                    it.remove();
            }
            return false;
        });

        //Then
        assertListsSame(expList, actList, new ChangeListsHelper.SimpleIntPair[0]);
    }

    @Test(timeout = 100)
    public void removeByListIterator() {
        //Given
        List<ChangeListsHelper.SimpleIntPair> originalList = Arrays.asList(
                new ChangeListsHelper.SimpleIntPair(1, 2),
                new ChangeListsHelper.SimpleIntPair(1, 3),
                new ChangeListsHelper.SimpleIntPair(1, 4),
                new ChangeListsHelper.SimpleIntPair(1, 5),
                new ChangeListsHelper.SimpleIntPair(2, 5),
                new ChangeListsHelper.SimpleIntPair(3, 5),
                new ChangeListsHelper.SimpleIntPair(4, 5)
        );
        List<ChangeListsHelper.SimpleIntPair> expList = new LinkedList<>(originalList);
        List<ChangeListsHelper.SimpleIntPair> actList = new SoftList<>(originalList);

        //When
        changeForEachListsSame(Arrays.asList(expList, actList), list -> {
            Iterator<ChangeListsHelper.SimpleIntPair> it = list.listIterator();
            while (it.hasNext()) {
                ChangeListsHelper.SimpleIntPair item = it.next();
                if (item.a >= 2)
                    it.remove();
            }
            return false;
        });

        //Then
        assertListsSame(expList, actList, new ChangeListsHelper.SimpleIntPair[0]);
    }


    @Test(timeout = 100)
    public void addByListIterator() {
        //Given
        List<ChangeListsHelper.SimpleIntPair> originalList = Arrays.asList(
                new ChangeListsHelper.SimpleIntPair(1, 2),
                new ChangeListsHelper.SimpleIntPair(1, 3)
        );
        List<ChangeListsHelper.SimpleIntPair> expList = new LinkedList<>(originalList);
        List<ChangeListsHelper.SimpleIntPair> actList = new SoftList<>(originalList);

        //When
        changeForEachListsSame(Arrays.asList(expList, actList), list -> {
            ListIterator<ChangeListsHelper.SimpleIntPair> it = list.listIterator();

            it.add(new ChangeListsHelper.SimpleIntPair(5, 5));
            return null;
        });

        //Then
        assertListsSame(expList, actList, new ChangeListsHelper.SimpleIntPair[0]);
    }


    @Test(timeout = 100)
    public void setByListIterator() {
        //Given
        List<ChangeListsHelper.SimpleIntPair> originalList = Arrays.asList(
                new ChangeListsHelper.SimpleIntPair(1, 2),
                new ChangeListsHelper.SimpleIntPair(1, 3)
        );
        List<ChangeListsHelper.SimpleIntPair> expList = new LinkedList<>(originalList);
        List<ChangeListsHelper.SimpleIntPair> actList = new SoftList<>(originalList);

        //When
        changeForEachListsSame(Arrays.asList(expList, actList), list -> {
            ListIterator<ChangeListsHelper.SimpleIntPair> it = list.listIterator();
            it.next();
            it.set(new ChangeListsHelper.SimpleIntPair(5, 5));
            return null;
        });

        //Then
        assertListsSame(expList, actList, new ChangeListsHelper.SimpleIntPair[0]);
    }

}

File: hummus/src/commonTest/kotlin/com/github/klee0kai/hummus/collections/weaklist/RefListIteratorTests.kt
package com.github.klee0kai.hummus.collections.weaklist

import com.github.klee0kai.hummus.collections.weaklist.ChangeListsHelper.assertListsSame
import com.github.klee0kai.hummus.collections.weaklist.ChangeListsHelper.changeForEachListsSame
import com.github.klee0kai.hummus.collections.weaklist.SoftList
import org.junit.Test
import java.util.*

class RefListIteratorTests {

    @Test(timeout = 100)
    fun removeByIterator() {
        //Given
        val originalList = listOf(
            Pair(1, 2),
            Pair(1, 3),
            Pair(1, 4),
            Pair(1, 5),
            Pair(2, 5),
            Pair(3, 5),
            Pair(4, 5)
        )
        val expList = LinkedList(originalList)
        val actList = SoftList(originalList)

        //When
        changeForEachListsSame(expList, actList) {
            val it = iterator()
            while (it.hasNext()) {
                val item = it.next()
                if (item.first >= 2) it.remove()
            }

        }

        //Then
        assertListsSame(expList, actList)
    }

    @Test(timeout = 100)
    fun removeByListIterator() {
        //Given
        val originalList = listOf(
            Pair(1, 2),
            Pair(1, 3),
            Pair(1, 4),
            Pair(1, 5),
            Pair(2, 5),
            Pair(3, 5),
            Pair(4, 5)
        )
        val expList = LinkedList(originalList)
        val actList = SoftList(originalList)

        //When
        changeForEachListsSame(expList, actList) {
            val it = listIterator()
            while (it.hasNext()) {
                val item = it.next()
                if (item.first >= 2) it.remove()
            }

        }

        //Then
        assertListsSame(expList, actList)
    }


    @Test(timeout = 100)
    fun addByListIterator() {
        //Given
        val originalList = listOf(
            Pair(1, 2),
            Pair(1, 3)
        )
        val expList = LinkedList(originalList)
        val actList = SoftList(originalList)

        //When
        changeForEachListsSame(expList, actList) {
            val it = listIterator()
            it.add(Pair(5, 5))
        }

        //Then
        assertListsSame(expList, actList)
    }


    @Test(timeout = 100)
    fun setByListIterator() {
        //Given
        val originalList = listOf(
            Pair(1, 2),
            Pair(1, 3)
        )

        val expList = LinkedList(originalList)
        val actList = SoftList(originalList)

        //When
        changeForEachListsSame(expList, actList) {
            val it = listIterator()
            it.next()
            it.set(Pair(5, 5))
        }

        //Then
        assertListsSame(expList, actList)
    }


}